application/src/main/java/org/togetherjava/tjbot/commands/free/AutoFreeRoutine.java

```diff
@@ -0,0 +1,72 @@
+package org.togetherjava.tjbot.commands.free;
+
+import net.dv8tion.jda.api.JDA;
+import net.dv8tion.jda.api.entities.Guild;
+import net.dv8tion.jda.api.entities.MessageChannel;
+import net.dv8tion.jda.api.entities.TextChannel;
+import net.dv8tion.jda.api.utils.TimeUtil;
+import org.jetbrains.annotations.NotNull;
+import org.togetherjava.tjbot.commands.Routine;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Collection;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * Routine that automatically marks busy help channels free after a certain time without any
+ * activity.
+ */
+public final class AutoFreeRoutine implements Routine {
+    private static final Duration INACTIVE_CHANNEL_DURATION = Duration.ofHours(2);
+    private final FreeChannelMonitor channelMonitor;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param channelMonitor used to monitor and control the free-status of channels
+     */
+    public AutoFreeRoutine(@NotNull FreeChannelMonitor channelMonitor) {
+        this.channelMonitor = channelMonitor;
+    }
+
+    @Override
+    public void runRoutine(@NotNull JDA jda) {
+        Instant now = Instant.now();
+
+        channelMonitor.getBusyChannelIds()
+            .mapToObj(jda::getTextChannelById)
+            .filter(busyChannel -> isInactive(busyChannel, now))
+            .collect(Collectors.groupingBy(TextChannel::getGuild))
+            .forEach(this::processChannelsForGuild);
+    }
+
+    private void processChannelsForGuild(@NotNull Guild guild,
+            @NotNull Collection<? extends TextChannel> inactiveChannels) {
+        // Mark all of them free
+        inactiveChannels.stream()
+            .mapToLong(TextChannel::getIdLong)
+            .forEach(channelMonitor::setChannelFree);
+
+        // Then update the status
+        channelMonitor.displayStatus(guild);
+
+        // Finally, send the messages (the order is important to ensure sane behavior in case of
+        // crashes)
+        inactiveChannels.forEach(inactiveChannel -> inactiveChannel
+            .sendMessage(UserStrings.AUTO_MARK_AS_FREE.message())
+            .queue());
+    }
+
+    private static boolean isInactive(@NotNull MessageChannel channel, @NotNull Instant when) {
+        return TimeUtil.getTimeCreated(channel.getLatestMessageIdLong())
+            .toInstant()
+            .isBefore(when.minus(INACTIVE_CHANNEL_DURATION));
+    }
```

Comment:
Why not just FreeChannelMonitor#updateStatusFor?

Reply:
Not sure what you mean. I wanted to process the channels grouped by guild to prevent issues. For example, it is more important that the channels are marked free and that the status message is updated than to send the actual "channel is free" message.

Because of that, I split those steps. So if the last step (sending messages) fails for some reason, the system is at least in a sane state.

Reply:
Also, I did not want other guilds to interfer with the process. If a guild fails, it shouldnt influence another guild.

Reply:
Dont forget that `channelMonitor.displayStatus(guild);` is an operation that should happen once per guild, not per channel. The rest of the operations are indeed per-channel only, but this one is per-guild.

Reply:
Once I get home I'll check.

But isn't this code just a very long & selfmade version of updateStatusFor + forEach of all guilds?

Pretty sure updateStatusFor checks each channel, once inactive sets it to free/available?

Reply:
oh... for fucks sake. All the logic was already implemented as dead-code... I just saw this lmao... 😆 FFS

Reply:
If you read this issue you saw that borgrel made a comment about it lol

Reply:
yeah i know, i didnt read it 😆

Reply:
done. please check again.